### PR TITLE
Midi support for Entities

### PR DESCRIPTION
**Pull request by burtsloane**
Entities can receive incoming MIDI notes from any device, and can send notes to all devices. USB plug/unplug is detected so that the system can re-adjust live. Device names are available and can be used to ignore specific input or output devices.

Sample scripts: createMidiSphere.js creates an Entity running midiSphere.js, which is a sphere that changes color when incoming notes are seen, and sends notes when clicked.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Can one of the admins verify this patch?

**Comment by birarda** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/772/HighFidelity-Beta-PR10711-2358d51ef14bd94dfe41356e194f9945b9da4bb4.exe

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -518,8 +523,9 @@ bool setupEssentials(int& argc, char** argv, bool runningMarkerExisted) {
     DependencyManager::set<DiscoverabilityManager>();
     DependencyManager::set<SceneScriptingInterface>();
     DependencyManager::set<OffscreenUi>();
-    DependencyManager::set<AutoUpdater>();
-    DependencyManager::set<PathUtils>();
+	DependencyManager::set<AutoUpdater>();
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: formatting is busted

File: interface/src/Application.cpp [hunk 2]
```diff
@@ -640,7 +646,7 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
     _entityClipboard->createRootElement();
 
 #ifdef Q_OS_WIN
-    installNativeEventFilter(&MyNativeEventFilter::getInstance());
+	installNativeEventFilter(&MyNativeEventFilter::getInstance());
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: formatting is busted

File: libraries/midi/src/Midi.cpp [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+
+void CALLBACK MidiInProc(
+	HMIDIIN   hMidiIn,
+	UINT      wMsg,
+	DWORD_PTR dwInstance,
+	DWORD_PTR dwParam1,
+	DWORD_PTR dwParam2
+) {
+	if (wMsg == MIM_OPEN) {
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: formatting is busted

File: libraries/midi/src/Midi.cpp [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	}
+	else if (wMsg == MIM_CLOSE) {
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard: `} else if (...) {`

File: libraries/midi/src/Midi.cpp [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+		for (int i = 0; i < midihin.size(); i++) if (midihin[i] == hMidiIn) {
+			midihin[i] = NULL;
+			instance->sendNote(0xb0, 0x7b, 0);		// all notes off
+		}
+	}
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard

File: libraries/midi/src/Midi.cpp [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+
+void Midi::sendNote(int status, int note, int vel) {
+	for (int i = 0; i < midihout.size(); i++) if (midihout[i] != NULL) {
+		midiOutShortMsg(midihout[i], status + (note << 8) + (vel << 16));
+	}
+}
+
+void Midi::noteReceived(int status, int note, int velocity) {
+	if (status >= 0x0a0) return;			// NOTE: only sending note-on and note-off to Javascript
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard

File: libraries/midi/src/Midi.cpp [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	midihin.clear();
+	midihout.clear();
+
+	MIDIINCAPS incaps;
+	for (unsigned int i = 0; i < midiInGetNumDevs(); i++) {
+		midiInGetDevCaps(i, &incaps, sizeof(MIDIINCAPS));
+
+		bool found = false;
+		for (int j = 0; j < midiinexclude.size(); j++) {
+			if (lstrcmp(midiinexclude[j].toStdString().c_str(), incaps.szPname) == 0) found = true;
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard

File: libraries/midi/src/Midi.cpp [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+		}
+		if (!found)			// EXCLUDE AN INPUT BY NAME
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard

**Comment by burtsloane** [comment]
Many fixes in place for indents, bracket placement, single-line if's

**Comment by burtsloane** [comment]
@ZappoMan I have addressed the code formatting and other issues. Would you please have another look?

File: libraries/midi/src/Midi.cpp [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+#pragma comment(lib, "Winmm.lib")
+
+//
+std::vector<HMIDIIN> midihin;
+std::vector<HMIDIOUT> midihout;
+
+
+void CALLBACK MidiInProc(HMIDIIN hMidiIn, UINT wMsg, DWORD_PTR dwInstance, DWORD_PTR dwParam1, DWORD_PTR dwParam2) {
+    if (wMsg == MIM_OPEN) {
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: this is weird, why is this empty clause here? maybe this should be a switch statement instead?

File: libraries/midi/src/Midi.cpp [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            if (midihin[i] == hMidiIn) {
+                midihin[i] = NULL;
+                instance->allNotesOff();
+            }
+        }
+    } else if (wMsg == MIM_DATA) {
+        int status = 0x0ff & dwParam1;
+        int note = 0x0ff & (dwParam1 >> 8);
+        int vel = 0x0ff & (dwParam1 >> 16);
+//sendNote(status, note, vel);        // NOTE: relay the note on to all other midi devices
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: remove this dead code and/or fix the formatting why is this outdented?

File: libraries/midi/src/Midi.cpp [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+            }
+        }
+    }
+}
+
+
+void Midi::sendNote(int status, int note, int vel) {
+    for (int i = 0; i < midihout.size(); i++) {
+        if (midihout[i] != NULL) {
+            midiOutShortMsg(midihout[i], status + (note << 8) + (vel << 16));
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard: no magic numbers, use named constants for whatever the 8 and 16 shifts are here... presumably they are part of the midi sdk?

File: libraries/midi/src/Midi.cpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+    } else if (wMsg == MIM_CLOSE) {
+        for (int i = 0; i < midihin.size(); i++) {
+            if (midihin[i] == hMidiIn) {
+                midihin[i] = NULL;
+                instance->allNotesOff();
+            }
+        }
+    } else if (wMsg == MIM_DATA) {
+        int status = 0x0ff & dwParam1;
+        int note = 0x0ff & (dwParam1 >> 8);
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard: no magic numbers, use named constants for whatever the 8 and 16 shifts are here... presumably they are part of the midi sdk?

File: libraries/midi/src/Midi.cpp [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    for (int i = 0; i < midihout.size(); i++) {
+        if (midihout[i] != NULL) {
+            midiOutShortMsg(midihout[i], status + (note << 8) + (vel << 16));
+        }
+    }
+}
+
+void Midi::noteReceived(int status, int note, int velocity) {
+    if (((status & MIDI_STATUS_MASK) != MIDI_NOTE_OFF) &&
+        ((status & MIDI_STATUS_MASK) != MIDI_NOTE_ON)) return;            // NOTE: only sending note-on and note-off to Javascript
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard... you still need to add the braces here

File: libraries/midi/src/Midi.cpp [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+    for (unsigned int i = 0; i < midiInGetNumDevs(); i++) {
+        midiInGetDevCaps(i, &incaps, sizeof(MIDIINCAPS));
+
+        bool found = false;
+        for (int j = 0; j < midiinexclude.size(); j++) {
+            if (midiinexclude[j].toStdString().compare(incaps.szPname) == 0) {
+                found = true;
+            }
+        }
+        if (!found)    {        // EXCLUDE AN INPUT BY NAME
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: coding standard: no large extra space between `)` and `{`

File: libraries/midi/src/Midi.cpp [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    MIDIOUTCAPS outcaps;
+    for (unsigned int i = 0; i < midiOutGetNumDevs(); i++) {
+        midiOutGetDevCaps(i, &outcaps, sizeof(MIDIINCAPS));
+
+        bool found = false;
+        for (int j = 0; j < midioutexclude.size(); j++) {
+            if (midioutexclude[j].toStdString().compare(outcaps.szPname) == 0) {
+                found = true;
```
Comment (ZappoMan) [inline libraries/midi/src/Midi.cpp]: do you want to early exit here?

File: libraries/midi/src/Midi.cpp (hunk 12 above)
Comment (burtsloane) [inline libraries/midi/src/Midi.cpp]: In fact, the shift amounts are not part of the SDK, and are described very oddly in the window docs
https://msdn.microsoft.com/en-us/library/windows/desktop/dd798481(v=vs.85).aspx

**Comment by burtsloane** [comment]
@ZappoMan thanks for your eval, I've fixed the above issues

**Comment by RyanDowne** [comment]
@burtsloane I am trying to test this but can't get it to work. Most likely my user error. I download the client, have bitwig running and confirm that I am getting a midi signal to it, it triggers a kick drum sound, run the script in your build and see the sphere. When I trigger a sound from the keyboard, the sphere color does not change. When I mouse click on the sphere, I do not hear the sound trigger. What am I doing wrong?

**Comment by burtsloane** [comment]
Hi Ryan-

It's not user error, it's me. I didn't anticipate that an app on the
machine could lock up the MIDI interfaces.

I downloaded BitWig, and once it was dialed in, it grabbed exclusive
control of the inputs from my two keyboards. When I ran my test harness (a
simple note repeater) for the windows-specific code, the trace says it
could not open the devices, so no sound. When I quit BitWig and restarted
my test harness, I was able to send a MIDI note from one keyboard to the
other.

In the case of the Midi library, there's a JS visible call to re-scan the
MIDI, which would avoid the need to restart. Also, for testing, a call to
Midi.thruModeEnable(true) would enable a simple repeater that takes any
incoming MIDI note and pumps it out to all.

Hope that helps,
- Burt

**Comment by ZappoMan** [comment]
@burtsloane - are you saying you can/should update the code to check for devices becoming available after startup of the script?

**Comment by burtsloane** [comment]
Well, it's not clear what I can do. There is a check if a USB device is
plugged or unplugged, which re-scans the bus, supporting clean plug-n-play
while the script is running. However, I'm not sure how to detect if a DAW
application on the same PC has recently quit and closed some MIDI devices,
releasing them for use. The worst case here is that the Midi library won't
know about these recently released MIDI devices until the app is restarted
(or someone calls Midi.resetDevices() or a windows event calls
Midi.USBchanged()).

A quick check with BitWig shows they can't detect this event either. I
started my test program, which opened all the MIDI devices it could find,
apparently locking them from other apps. I then started BitWig, and even
though it didn't complain, and showed it had connections, no notes would
pass. I then quit my test harness, which closed all the MIDI devices making
them available, and BitWig did not notice, still no notes. Not until I
restarted BitWig was it able to hear incoming MIDI notes.

Cheers!
- Burt

**Comment by RyanDowne** [comment]
Thanks, Burt. I am still trying to test this somehow. For us non-technical folk, PR's usually include a test plan for us to follow. Would you be willing to provide that? 

Right now I have a midi keyboard plugged into the PC, turned on and follow your instructions above - creating the sphere that has the script running. When I press my keyboard notes I don't see the sphere change color. I do not have a DAW running. Perhaps I don't have the right midi drivers on my PC?

**Comment by burtsloane** [comment]
You have the test plan correct, but I will add it to the transcript.

Would you be so kind as to test this app, which should indicate any
problems? If you could run it and grab a screen shot, I might be able to
diagnose. I had to rename the extension to .xe to get through gmail, please
rename it back to .exe and double-click.

- Burt

**Comment by burtsloane** [comment]
TEST PLAN

1. Start the createMidiSphere.js script to create the sphere
2. Click on the sphere, it should change color and send out a note to all connected MIDI devices
3. Click again to turn off the note
4. Play a note on any connected MIDI device, the velocity of the note is mapped to the grey level of the sphere, dark for low velocity, light for high

**Comment by RyanDowne** [comment]
@burtsloane I don't have access to the email folder your email replies go to. I think you attached your test app via email in your response. Can you email that directly to me - ryan@highfidelity.io 

thanks

File: libraries/midi/src/Midi.cpp (hunk 15 above)
Comment (burtsloane) [inline libraries/midi/src/Midi.cpp]: This has been fixed

**Comment by RyanDowne** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1212/HighFidelity-Beta-PR10711-f1485afccc80585499dd0d6b9ab4691f74f85853.exe

**Comment by RyanDowne** [comment]
I get this error in the log when trying to load the createMidisphere.js

[06/29 13:29:50] [CRITICAL] [hifi.scriptengine] [UncaughtException evaluate(lint)] Error: Expected `end of file' in file:///C:/Program Files/High Fidelity - PR10711/scripts/tutorials/createMidiSphere.js:13

**Comment by 1st-BrainStormer** [comment]
After removing the extra quote on line 12 of MidiSphere.js Test was sucessfull.
Sphere was created.
Clicking on the sphere played a note.
Clicking again turned off note.
Playing a note changes shade on the first hit but subsequent hits don't until a different note is played.

**Comment by burtsloane** [comment]
There was, somehow, an extra quote in the file. This has been removed.

**Comment by cozza13** [comment]
OK to test

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1272/HighFidelity-Beta-PR10711-2b111358e1cb445d14128c00c7e1712986c5b0b2.exe

**Comment by 1st-BrainStormer** [comment]
Sphere was created.
Clicking on the sphere played a note.
Clicking again turned off note.
Playing a note changes shade on the first hit but subsequent hits don't until a different note is played.

SUCCESS

Note:  did have my Midi driver crash and lockup interface.  My USB extension cable is bad and caused errors on the usb bus causing the midi out to disconnect.

**Comment by themelissabrown** [comment]
@burtsloane looks like there are a few changes requested

**Comment by burtsloane** [comment]
The requested change mentioned above has been made. I don't have any information about the check that failed (no permissions)

**Comment by burtsloane** [comment]
Not sure how requested changes are checked, but I've made the one that looks open, a while ago

**Comment by themelissabrown** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1551/HighFidelity-Beta-PR10711-2b111358e1cb445d14128c00c7e1712986c5b0b2.exe

**Comment by 1st-BrainStormer** [comment]
TEST PLAN

1. Start the createMidiSphere.js script to create the sphere
    Success.
2. Click on the sphere, it should change color and send out a note to all connected MIDI devices
   Success.
3. Click again to turn off the note
    Success.
4. Play a note on any connected MIDI device, the note is mapped to the grey level of the sphere, dark for lower notes, light for higher.  (No longer Velocity)
    Success.

**Comment by themelissabrown** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1687/HighFidelity-Beta-PR10711-2b111358e1cb445d14128c00c7e1712986c5b0b2.exe

**Comment by 1st-BrainStormer** [comment]
Still works fine

**Comment by 1st-BrainStormer** [comment]
Whats holding this up?  I would like to get my Novation Launchpad Pro working with my virtual one. :-)

**Comment by burtsloane** [comment]
Wish I knew!
Could be because there was a code change requested by ZappoMan 6/16 that was sticky and didn't get approved till 7/11. Also, Jenkins is reporting "All checks have failed". I'm unable to drill down due to lack of access. 
What would you expect from a VR Launchpad connected to a real one? Echos, or more like the ganging you can see in https://www.youtube.com/watch?v=zGEXqHyjJeE

**Comment by AndrewMeadows** [comment]
Jenkins, rebuild this please.

**Comment by AndrewMeadows** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1934/HighFidelity-Beta-PR10711-2b111358e1cb445d14128c00c7e1712986c5b0b2.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1972/HighFidelity-Beta-PR10711-88838da6e584dd2375909173dd3da250c494e932.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1977/HighFidelity-Beta-PR10711-246a5c9fa4ca91cfa76c7bbaa7fb093395c1031c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1977/HighFidelity-Beta-PR10711-246a5c9fa4ca91cfa76c7bbaa7fb093395c1031c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1981/HighFidelity-Beta-PR10711-c777384d939ce8c43a67dfd1473f11ef95d6850b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1981/HighFidelity-Beta-PR10711-c777384d939ce8c43a67dfd1473f11ef95d6850b.exe

**Comment by AndrewMeadows** [comment]
@burtsloane, I cloned your branch on linux and tried to build it.  The diff below would allow the build to succeed.  In short: **(a)** the `script-engine` library depends on `midi` and **(b)** there is no need to add `midi` dependency to the `test/controllers` project.

Also note: there are conflicts that would prevent us from merging this PR.  You need to perform yet another pull from hifi/master and resolve the conflicts.
```
diff --git a/libraries/script-engine/CMakeLists.txt b/libraries/script-engine/CMakeLists.txt                                                                                                                                                  
index 39338fd..42b2d51 100644                                                                                                                                                                                                                 
--- a/libraries/script-engine/CMakeLists.txt                                                                                                                                                                                                  
+++ b/libraries/script-engine/CMakeLists.txt                                                                                                                                                                                                  
@@ -16,6 +16,6 @@ if (NOT ANDROID)                                                                                                                                                                                                            
 
 endif ()
 
-link_hifi_libraries(shared networking octree gpu ui procedural model model-networking ktx recording avatars fbx entities controllers animation audio physics image)
+link_hifi_libraries(shared networking octree gpu ui procedural model model-networking ktx recording avatars fbx entities controllers animation audio physics image midi)
 # ui includes gl, but link_hifi_libraries does not use transitive includes, so gl must be explicit
 include_hifi_library_headers(gl)
diff --git a/tests/controllers/CMakeLists.txt b/tests/controllers/CMakeLists.txt
index 7be7192..5757a3a 100644
--- a/tests/controllers/CMakeLists.txt
+++ b/tests/controllers/CMakeLists.txt
@@ -6,7 +6,7 @@ setup_hifi_project(Script Qml)
 set_target_properties(${TARGET_NAME} PROPERTIES FOLDER "Tests/manual-tests/")
 
 # link in the shared libraries
-link_hifi_libraries(shared gl script-engine plugins render-utils ui-plugins input-plugins display-plugins controllers midi)
+link_hifi_libraries(shared gl script-engine plugins render-utils ui-plugins input-plugins display-plugins controllers)
 
 
 if (WIN32)
```

**Comment by burtsloane** [comment]
Thanks- I've made the changes, and confirmed that I can build on OS X. I
don't have access to a Linux box, so thanks again.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1989/HighFidelity-Beta-PR10711-a960ea5ccbba13fbeba68c55fcbd3de43bcb46c0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/1989/HighFidelity-Beta-PR10711-a960ea5ccbba13fbeba68c55fcbd3de43bcb46c0.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/1991/HighFidelity-Beta-PR10711-d4ae298a8d9533c6dff1e8b24f6fff3b41347e4e.dmg

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2000/HighFidelity-Beta-PR10711-d4ae298a8d9533c6dff1e8b24f6fff3b41347e4e.dmg

**Comment by marko8904** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2027/HighFidelity-Beta-PR10711-d4ae298a8d9533c6dff1e8b24f6fff3b41347e4e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows2016/2027/HighFidelity-Beta-PR10711-d4ae298a8d9533c6dff1e8b24f6fff3b41347e4e.exe

**Comment by AndrewMeadows** [comment]
I'm sorry @burtsloane, the Mac build is failing for three `const` variables that are declared but never used:
```
const int MIDI_BYTE_MASK = 0x0FF;
const int MIDI_SHIFT_NOTE = 8;
const int MIDI_SHIFT_VELOCITY = 16;
```
You'll have to remove those lines.  I didn't check the Mac build failure earlier because I assumed it was failing for the same reason as the linux one.

BTW, you should be able to figure out the build failure mode by following the **Details** link and then clicking on the red operating system links.

**Comment by burtsloane** [comment]
Thanks for tracking this down. I've made the changes. BTW, I have tried the
"Details" link, but here's what I get:
Access Denied
burtsloane is missing the Overall/Read permission

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/2062/HighFidelity-Beta-PR10711-e557f759b592468320db14231953c88c47baf5d7.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/2062/HighFidelity-Beta-PR10711-e557f759b592468320db14231953c88c47baf5d7.exe

**Comment by 1st-BrainStormer** [comment]
Cool. here are some of the scrips I have been working on:

Rezes in-world midi keyboard or launchpad depending on wantKeyboard var
http://brainstormer.ddns.net:650/Entities/Launchpad/Launchpad.js
___________________________________________________________________
Demonstrates Midi
http://brainstormer.ddns.net:650/Entities/Launchpad/midiTest.js
___________________________________________________________________
Use midi to trigger sounds and animations, (Also lights the Novation Launchpad Pro keys)
http://brainstormer.ddns.net:650/Entities/Launchpad/midiEmote.js

Not sure if there is a way to use midi control events or sysex commands yet so I can better control my launchpad.

**Pull request merged**